x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/LongTopNBlockHash.java

```diff
@@ -0,0 +1,311 @@
+            } else {
+                topValues.remove(topValues.last());
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns true if the value is in, or can be added to the top; false otherwise.
+     */
+    private boolean isAcceptable(long value) {
+        return isTopComplete() == false || (hasNull && nullsFirst == false) || isInTop(value);
+    }
+
+    /**
+     * Returns true if the value is in the top; false otherwise.
+     * <p>
+     *     This method does not check if the value is currently part of the top; only if it's better or equal than the current worst value.
+     * </p>
+     */
+    private boolean isInTop(long value) {
+        return asc ? value <= topValues.last() : value >= topValues.last();
+    }
+
+    /**
+     * Returns true if there are {@code limit} values in the blockhash; false otherwise.
+     */
+    private boolean isTopComplete() {
+        return topValues.size() >= limit - (hasNull ? 1 : 0);
+    }
+
+    /**
+     *  Adds the vector values to the hash, and returns a new vector with the group IDs for those positions.
+     */
+    IntBlock add(LongVector vector) {
+        int positions = vector.getPositionCount();
+
+        // Add all values to the top set, so we don't end up sending invalid values later
+        for (int i = 0; i < positions; i++) {
+            long v = vector.getLong(i);
+            acceptValue(v);
+        }
+
+        // Create a vector with the groups
+        try (var builder = blockFactory.newIntBlockBuilder(positions)) {
+            for (int i = 0; i < positions; i++) {
+                long v = vector.getLong(i);
+                if (isAcceptable(v)) {
+                    builder.appendInt(Math.toIntExact(hashOrdToGroupNullReserved(hash.add(v))));
+                } else {
+                    builder.appendNull();
+                }
+            }
+            return builder.build();
+        }
+    }
+
+    /**
+     *  Adds the block values to the hash, and returns a new vector with the group IDs for those positions.
+     * <p>
+     *     For nulls, a 0 group ID is used. For multivalues, a multivalue is used with all the group IDs.
+     * </p>
+     */
+    IntBlock add(LongBlock block) {
+        // Add all the values to the top set, so we don't end up sending invalid values later
+        for (int p = 0; p < block.getPositionCount(); p++) {
+            int count = block.getValueCount(p);
+            if (count == 0) {
+                acceptNull();
+                continue;
+            }
+            int first = block.getFirstValueIndex(p);
+
+            for (int i = 0; i < count; i++) {
+                long value = block.getLong(first + i);
+                acceptValue(value);
+            }
+        }
+
+        // TODO: Make the custom dedupe *less custom*
+        MultivalueDedupe.HashResult result = new TopNMultivalueDedupeLong(block, hasNull, this::isAcceptable).hashAdd(blockFactory, hash);
```

Comment:
`MultivalueDedupe` might be what you need.

Reply:
Adding here: This `TopNMultivalueDedupeLong` is currently a plain copy of that class, with the methods slightly changed to work with the TopN logic.
I'll see if I can merge it into the original one. Otherwise, I may trim the unused parts, and leave it as a new class. Or a nested class